nexus/src/app/background/tasks/multicast/members.rs

```diff
@@ -0,0 +1,1495 @@
+        Err(anyhow::Error::msg(format!(
+            "failed to resolve sled to switch ports: \
+             sled {sled_id} not found in mapping cache (not a scrimlet or removed)"
+        )))
+    }
+
+    /// Refresh the sled-to-switch-port mapping cache.
+    async fn refresh_sled_mapping_cache(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<(), anyhow::Error> {
+        // Get all scrimlets (switch-connected sleds) from the database
+        let sleds = self
+            .datastore
+            .sled_list_all_batched(
+                opctx,
+                nexus_types::deployment::SledFilter::Commissioned,
+            )
+            .await
+            .context("failed to list sleds")?;
+
+        // Filter to only scrimlets
+        let scrimlets: Vec<_> =
+            sleds.into_iter().filter(|sled| sled.is_scrimlet()).collect();
+
+        trace!(
+            opctx.log,
+            "building sled mapping cache for scrimlets";
+            "scrimlet_count" => scrimlets.len()
+        );
+
+        let mut mappings = HashMap::new();
+
+        // For each scrimlet, determine its switch location from switch port data
+        for sled in scrimlets {
+            // Query switch ports to find which switch this sled is associated with
+            // In the Oxide rack, each scrimlet has a co-located switch
+            // We need to find switch ports that correspond to this sled's location
+            let switch_ports = self
+                .datastore
+                .switch_port_list(opctx, &DataPageParams::max_page())
+                .await
+                .context("failed to list switch ports")?;
+
+            // Find ports that map to this scrimlet
+            let instance_switch_ports = match self
+                .find_instance_switch_ports_for_sled(&sled, &switch_ports)
+            {
+                Some(ports) => ports,
+                None => {
+                    return Err(anyhow::Error::msg(format!(
+                        "no instance switch ports found for sled {} - cannot create multicast mapping (sled rack_id: {})",
+                        sled.id(),
+                        sled.rack_id
+                    )));
+                }
+            };
+
+            // Create mappings for all available instance ports on this sled
+            let mut sled_port_configs = Vec::new();
+            for instance_switch_port in instance_switch_ports.iter() {
+                // Set port and link IDs
+                let port_id = instance_switch_port
+                    .port_name
+                    .as_str()
+                    .parse()
+                    .context("failed to parse port name")?;
+                let link_id = dpd_client::types::LinkId(0);
+
+                let config = MulticastSwitchPort {
+                    port_id,
+                    link_id,
+                    direction: dpd_client::types::Direction::Underlay,
```

Comment:
I'm a bit confused about what's going on here. The `find_instance_switch_ports_for_sled` code above looks like it returns all QSFP (external facing ports) for a scrimlet. But the direction here is `Underlay` which is the opposite of what I would expect. A sled-to-switch-port mapping also leads me to believe that the intent is to determine the switch backplane ports (not QSFP) for a given sled, but that does not appear to be what's going on here.

Reply:
Coming back to this in the next commit.

Reply:
Thanks, the changes in 38d9a18. Look much better. I read through the implementation and have a few questions around cache invalidation.